### PR TITLE
20220817 ldap service tokens

### LINKED ISSUE #466: Schema should allow class depenedncies
Today we have a few types like posixaccount or acp types that don't make sense *without* another supplemental class. We should have a way to express that these require something to be valid. Unlikel a class inheritance, this should simply be "this class to be valid, means the object must ALSO have this class present". Alternately, we could make it "Must have one of these classes" to allow a situation where posixAccount could be a valid extension to multiple possible basetypes for example.

### LINKED ISSUE #655: CA trust doesn't seem to work while running kanidm
I'm trying to sort out the dev docs and ... the CA trust option isn't working?

```
➜ RUST_LOG=DEBUG cargo run --bin kanidm -- login -H https://localhost:8443 -D anonymous -C /tmp/kanidm/ca.pem
    Finished dev [unoptimized + debuginfo] target(s) in 0.26s
     Running `target/debug/kanidm login -H 'https://localhost:8443' -D anonymous -C /tmp/kanidm/ca.pem`
2022-03-27T11:22:42.920742Z DEBUG kanidm_client: Attempting to load configuration from "/etc/kanidm/config"
2022-03-27T11:22:42.921004Z DEBUG kanidm_client: Configuration file "/etc/kanidm/config" not found, skipping.
2022-03-27T11:22:42.921029Z DEBUG kanidm_client: Attempting to load configuration from "/Users/yaleman/.config/kanidm"
2022-03-27T11:22:42.921071Z DEBUG kanidm_client: Configuration file "/Users/yaleman/.config/kanidm" not found, skipping.
2022-03-27T11:22:42.921113Z DEBUG kanidm_cli::common: Successfully read configuration from /Users/yaleman/.config/kanidm
2022-03-27T11:22:42.921426Z  WARN kanidm_client: permissions on /tmp/kanidm/ca.pem may not be secure. Should be readonly to running uid. This could be a security risk ...
2022-03-27T11:22:42.921449Z  WARN kanidm_client: /tmp/kanidm/ca.pem should be owned be root:root to prevent tampering
2022-03-27T11:22:42.932406Z DEBUG reqwest::connect: starting new connection: https://localhost:8443/
2022-03-27T11:22:42.932618Z DEBUG hyper::client::connect::dns: resolving host="localhost"
2022-03-27T11:22:42.933563Z DEBUG hyper::client::connect::http: connecting to [::1]:8443
2022-03-27T11:22:42.933768Z DEBUG hyper::client::connect::http: connecting to 127.0.0.1:8443
2022-03-27T11:22:42.933972Z DEBUG hyper::client::connect::http: connected to 127.0.0.1:8443
2022-03-27T11:22:42.949582Z ERROR kanidm_cli::session: Error during authentication init phase: Transport(reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(8443), path: "/v1/auth", query: None, fragment: None }, source: hyper::Error(Connect, Error { code: -67843, message: "The certificate was not trusted." }) })
```

Testing it with curl works fine so it's not the cert.

```
 ➜ curl -I --cacert /tmp/kanidm/ca.pem https://localhost:8443
HTTP/1.1 200 OK
cache-control: max-age=3600,private
content-length: 1024
content-type: text/html;charset=utf-8
cross-origin-embedder-policy: require-corp
cross-origin-opener-policy: same-origin
cross-origin-resource-policy: same-origin
date: Sun, 27 Mar 2022 11:22:32 GMT
set-cookie: kanidm-session=R9wllMxGtwI1ZonG7CUoIIhRNXw4zaUaayxJJP+5do0%3DlClpwuyzybuOpiuxQvDU+4NrA%2FyXgZAccs3kZL4PokoRhLOIoAGnfdxWZamKvLpXiRitIGT1e75LPYuY87Q35Q%3D%3D; HttpOnly; SameSite=Strict; Secure; Path=/; Expires=Mon, 28 Mar 2022 11:22:32 GMT
x-content-type-options: nosniff
x-frame-options: deny
```

### LINKED ISSUE #990: LDAP service accounts with additional permissions / Access mail attribute via LDAP for Postfix
**Is your feature request related to a problem? Please describe.**
I'm trying to deploy Kanidm in a self-hosting/hobbyist scenario with a handful of users, as a replacement for our exising FreeIPA installation. We run our own mail server with user authentication and mailbox mappings provided through LDAP. As the `mail` attribute in Kanidm is not readable by the anonymous user, LDAP lookups cannot read it.

**Describe the solution you'd like**
Since giving anonymous access to the `mail` attribute isn't exactly great from a security perspective, one option I've considered is the use of service accounts. Bind DNs are already usable for auth purposes in Kanidm, so maybe this could be extended to also include extra attributes the given user can read as long as the user has a "read_from_ldap" permission or something like that. This would also be useful for user federation when using IdPs like Keycloak.

**Describe alternatives you've considered**
Writing a Postfix plugin to use the Kanidm API directly? :skull:

### PR DESCRIPTION
**Pull request by Firstyear**
Partial #990 Fixes #655 Fixes #466 Partial #41. This starts to improve our handling of service accounts vs persons, by splitting them better away from the account with sub-person types, rather now having person and service accounts as the top level extended by account properties. This allows them to have seperated and exclusive features. Additionally this also splits the CLI and API to have seperate person and service account end points, with the shared end points being for combined functionality only. 

This also fixes the local CA generation to create valid CA certs. 

- [ x ] cargo fmt has been run
- [ ] cargo clippy has been run
- [ x ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

**Comment by Firstyear** [comment]
For once I delete more than I add!

**Comment by yaleman** [comment]
Shenanigans!

File: kanidm_book/src/accounts_and_groups.md [hunk 1]
```diff
@@ -3,92 +3,139 @@
 Accounts and Groups are the primary reasons for Kanidm to exist. Kanidm is optimised as a repository
 for these data. As a result, there are many concepts and important details to understand.
 
+## Service Accounts vs Person Accounts
+
+Kanidm seperates accounts into two types. Person accounts (or persons) are intended for use by humans
+that will access the system in an interactive way. Service accounts are intentded for use by computers
```
Comment (yaleman) [inline kanidm_book/src/accounts_and_groups.md:9]:
```suggestion
that will access the system in an interactive way. Service accounts are intended for use by computers
```

**Review by yaleman — CHANGES_REQUESTED** [review]
Some minor nits on the yak 😄

File: kanidm_book/src/posix_accounts.md [hunk 2]
```diff
@@ -72,19 +73,19 @@ capable of supplying this kind of data in batch jobs.
...
-    kanidm account posix set --name idm_admin demo_user --gidnumber 2001
+    kanidm person|service-account posix set --name idm_admin <account_id> [--shell SHELL --gidnumber GID]
+    kanidm person|service-account posix set --name idm_admin demo_user
+    kanidm person|service-account posix set --name idm_admin demo_user --shell /bin/zsh
+    kanidm person|service-account posix set --name idm_admin demo_user --gidnumber 2001
 
 You can view the accounts POSIX token details with:
 
-    kanidm account posix show --name anonymous demo_user
+    kanidm person|service-account posix show --name anonymous demo_user
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidm_book/src/posix_accounts.md:88]: Hmmm what would be a better way to say "or"? maybe two lines one for each? Means a lot of duplication though :(

**Comment by Firstyear** [comment]
Done, comments addressed :)

**Review by yaleman — APPROVED** [review]

**Pull request merged**